### PR TITLE
Introduces SchemaUtils to programatically drop and create schemas

### PR DESCRIPTION
**Pull request by octonato**
References #69

### CODE REVIEW

File: core/src/main/scala/akka/persistence/jdbc/schema/scaladsl/SchemaUtils.scala
Comment (ennru) [inline core/src/main/scala/akka/persistence/jdbc/schema/scaladsl/SchemaUtils.scala:61]: Make it compatible with the new actors API out-of-the-box:
```suggestion
  def dropAndCreate(implicit actorSystem: ClassicActorSystemProvider): Future[Done] = {
```

**Review by ennru — APPROVED** [review]
LGTM.

**Comment by octonato** [comment]
@ennru, a second review round is welcome. 

I did some more clean-ups, added Java API, API docs and updated the documentation.

File: docs/src/main/paradox/index.md
Comment (octonato) [inline docs/src/main/paradox/index.md:83]: we need to start using imports from code snippets, but I'm explicitly leaving that for another occasion so we can move faster with this PR.

File: core/src/main/scala/akka/persistence/jdbc/schema/scaladsl/SchemaUtils.scala
Comment (johanandren) [inline core/src/main/scala/akka/persistence/jdbc/schema/scaladsl/SchemaUtils.scala:44]: should it be `dropAndCreate()(implicits)` given that it is not a pure method, so that usage becomes `dropAndCreate()`?

File: core/src/main/scala/akka/persistence/jdbc/schema/scaladsl/SchemaUtils.scala
Comment (johanandren) [inline core/src/main/scala/akka/persistence/jdbc/schema/scaladsl/SchemaUtils.scala:79]: Default params are problematic for bincomp, not sure what we promise around that for akka-persistence-jdbc?

File: core/src/main/scala/akka/persistence/jdbc/schema/scaladsl/SchemaUtils.scala
Comment (johanandren) [inline core/src/main/scala/akka/persistence/jdbc/schema/scaladsl/SchemaUtils.scala:108]: Not worth logging at debug because we are so sure they are never interesting?

File: core/src/test/scala/akka/persistence/jdbc/journal/JdbcJournalPerfSpec.scala
Comment (johanandren) [inline core/src/test/scala/akka/persistence/jdbc/journal/JdbcJournalPerfSpec.scala:54]: .futureValue is available because `ScalaFutures` but perhaps we don't want the patience config for that here?

**Review by johanandren — APPROVED** [review]
LGTM, with some nitpicks

File: core/src/main/scala/akka/persistence/jdbc/schema/scaladsl/SchemaUtils.scala
Comment (ennru) [reply core/src/main/scala/akka/persistence/jdbc/schema/scaladsl/SchemaUtils.scala:79]: We do, this should be done with an overload instead.

File: core/src/main/scala/akka/persistence/jdbc/schema/scaladsl/SchemaUtils.scala
Comment (ennru) [inline core/src/main/scala/akka/persistence/jdbc/schema/scaladsl/SchemaUtils.scala:98]: Why is this wrapped in an option? How does it fail if the schema/filename can't be read as a resource?
Should there be an API to provide the DDL directly?

**Review by ennru — APPROVED** [review]
Looks still good!
A few suggestions.

File: core/src/main/resources/schema/oracle/oracle-schema.sql
Comment (ignasi35) [inline core/src/main/resources/schema/oracle/oracle-schema.sql:21]: I know this aligns the script to those for MySQL, Postgre, etc... where there is a DROP first and CREATE later, but this PR is also moving the `.sql` scripts to `main/` (from `test/`).

I'd rather we had (ideally idempotent) `.sql` scripts in `main/` that only created and then, somewhere else (e.g. `test/`) the scripts to drop.

Comment (ignasi35) [reply core/src/main/resources/schema/oracle/oracle-schema.sql:21]: As a side effect of that split, we would be improving some cases (for example postgre-schema.sql) where there is:

```sql
DROP TABLE IF EXISTS public.journal;

CREATE TABLE IF NOT EXISTS public.journal (
...
);

DROP TABLE IF EXISTS public.snapshot;

CREATE TABLE IF NOT EXISTS public.snapshot (
...
```

Which looks a bit weird ("create table if not exists" immediately after "drop table"?).

File: core/src/main/scala/akka/persistence/jdbc/schema/javadsl/SchemaUtils.scala
Comment (ignasi35) [inline core/src/main/scala/akka/persistence/jdbc/schema/javadsl/SchemaUtils.scala:27]: I think it'd be great to have a dedicated module for these tools. Getting them in production code, out of the box looks weird.

File: core/src/main/scala/akka/persistence/jdbc/schema/javadsl/SchemaUtils.scala
Comment (ignasi35) [inline core/src/main/scala/akka/persistence/jdbc/schema/javadsl/SchemaUtils.scala:21]:
```suggestion
object TestingUtils {
```

File: core/src/main/scala/akka/persistence/jdbc/schema/javadsl/SchemaUtils.scala
Comment (ignasi35) [inline core/src/main/scala/akka/persistence/jdbc/schema/javadsl/SchemaUtils.scala:6]:
```suggestion
package akka.persistence.jdbc.testkit.javadsl
```

File: core/src/main/scala/akka/persistence/jdbc/schema/scaladsl/SchemaUtils.scala
Comment (ignasi35) [inline core/src/main/scala/akka/persistence/jdbc/schema/scaladsl/SchemaUtils.scala:28]: Same rename suggestions.

File: core/src/test/scala/akka/persistence/jdbc/SingleActorSystemPerTestSpec.scala
Comment (ignasi35) [inline core/src/test/scala/akka/persistence/jdbc/SingleActorSystemPerTestSpec.scala:80]:
```suggestion
    SchemaUtils.dropAndCreate(system).futureValue
```

use `futureValue` from `org.scalatest.concurrent.ScalaFutures`?

Comment (ignasi35) [reply core/src/test/scala/akka/persistence/jdbc/SingleActorSystemPerTestSpec.scala:80]: Ah, just noticed @johanandren's comment about patience config.

**Review by ignasi35 — CHANGES_REQUESTED** [review]
I'd rather make it clearer the new `SchemaUtils` is not meant for production. Either renaming the object and/or the package, or even moving the code and SQL scripts into a new module users can import with scope `Test`.

**Comment by ignasi35** [comment]
> I'd rather make it clearer the new `SchemaUtils` is not meant for production. Either renaming the object and/or the package, or even moving the code and SQL scripts into a new module users can import with scope `Test`.

This suspiciously reminded me of https://github.com/lagom/lagom/blob/master/persistence-cassandra/javadsl/src/main/scala/com/lightbend/lagom/javadsl/persistence/cassandra/testkit/TestUtil.scala

File: core/src/main/scala/akka/persistence/jdbc/schema/javadsl/SchemaUtils.scala
Comment (patriknw) [reply core/src/main/scala/akka/persistence/jdbc/schema/javadsl/SchemaUtils.scala:27]: maybe a new module (jar artifact) is overkill, but just by placing them in a `testkit` package would clarify expectations

File: core/src/main/scala/akka/persistence/jdbc/schema/scaladsl/SchemaUtils.scala
Comment (patriknw) [reply core/src/main/scala/akka/persistence/jdbc/schema/scaladsl/SchemaUtils.scala:79]: I'd say that bincompat is a lost promise for this plugin until we rewrite it, but good to not introduce more warts

File: core/src/main/scala/akka/persistence/jdbc/schema/scaladsl/SchemaUtils.scala
Comment (patriknw) [reply core/src/main/scala/akka/persistence/jdbc/schema/scaladsl/SchemaUtils.scala:44]: why two so different names, `dropAndCreate` vs `applySchema` ?

File: core/src/test/scala/akka/persistence/jdbc/serialization/StoreOnlySerializableMessagesTest.scala
Comment (patriknw) [inline core/src/test/scala/akka/persistence/jdbc/serialization/StoreOnlySerializableMessagesTest.scala:59]: why so tight timeout?

**Review by patriknw — COMMENTED** [review]
looking good

File: core/src/main/scala/akka/persistence/jdbc/schema/scaladsl/SchemaUtils.scala
Comment (octonato) [reply core/src/main/scala/akka/persistence/jdbc/schema/scaladsl/SchemaUtils.scala:44]: > why two so different names, dropAndCreate vs applySchema ?

Because the default schema is a drop and create. 

The `applySchema` can be any DDL script.

Maybe better if we call if `applyDefaultSqlScript` and `applySqlScript`. 

The passed script can be anything you want. It can be an alter table, a drop, inserts.

File: core/src/main/scala/akka/persistence/jdbc/schema/scaladsl/SchemaUtils.scala
Comment (octonato) [reply core/src/main/scala/akka/persistence/jdbc/schema/scaladsl/SchemaUtils.scala:108]: Yes, will do it

File: core/src/test/scala/akka/persistence/jdbc/journal/JdbcJournalPerfSpec.scala
Comment (octonato) [reply core/src/test/scala/akka/persistence/jdbc/journal/JdbcJournalPerfSpec.scala:54]: that's how we did everywhere in akka-projections.

File: core/src/main/scala/akka/persistence/jdbc/schema/scaladsl/SchemaUtils.scala
Comment (octonato) [reply core/src/main/scala/akka/persistence/jdbc/schema/scaladsl/SchemaUtils.scala:30]: I was in doubt about this one. The javadsl relies on the scaladsl implementation. So, if you are using Java, you will have to configure the logging using `.scaladsl.`. 

Confusing, but also confusing the way I did. Maybe better to have a `.internal.SchemaUtils` to which we pass the logger from the Scala/Java one. So nobody gets surprised.

File: core/src/test/scala/akka/persistence/jdbc/serialization/StoreOnlySerializableMessagesTest.scala
Comment (octonato) [reply core/src/test/scala/akka/persistence/jdbc/serialization/StoreOnlySerializableMessagesTest.scala:59]: I'll can increase, but I find it high already.

File: core/src/main/scala/akka/persistence/jdbc/schema/javadsl/SchemaUtils.scala
Comment (octonato) [reply core/src/main/scala/akka/persistence/jdbc/schema/javadsl/SchemaUtils.scala:27]: I will move it to `testkit` although I don't think it's necessary. 

I don't think it's such a bad thing to not have it as is. We do the same in projection. The `createIfNotExists` methods are on the respective Projection companions. 

We have the scaladoc saying we don't recommend it. The method name is super explicit about it, it's `dropAndCreate`.

File: core/src/main/scala/akka/persistence/jdbc/schema/javadsl/SchemaUtils.scala
Comment (octonato) [reply core/src/main/scala/akka/persistence/jdbc/schema/javadsl/SchemaUtils.scala:21]: Not sure if we should to it. Isn't `testkit.SchemaUtils` good enough?

File: core/src/main/resources/schema/oracle/oracle-schema.sql
Comment (octonato) [reply core/src/main/resources/schema/oracle/oracle-schema.sql:21]: Ok to remove the `IF NOT EXISTS` from create table. 

Since the goal is to use it as a helper for building test, I do think it should drop and create by default. Users building tests do need to drop. If they don't want to drop the schema, they can use `applySchema` instead and pass a another file.

File: core/src/main/scala/akka/persistence/jdbc/schema/javadsl/SchemaUtils.scala
Comment (patriknw) [reply core/src/main/scala/akka/persistence/jdbc/schema/javadsl/SchemaUtils.scala:21]: `testkit` package is enough in my opinion

File: core/src/main/scala/akka/persistence/jdbc/schema/scaladsl/SchemaUtils.scala
Comment (octonato) [reply core/src/main/scala/akka/persistence/jdbc/schema/scaladsl/SchemaUtils.scala:98]: I copied this code from tests. I guess the reason is that we are flatMapping with another option, so we need to align. It could surely be different. 

> Should there be an API to provide the DDL directly?

I didn't want to increase the scope and the options. I found it was a good compromise to require the schema to be on the classpath, like in `src/test/resources`.

Can be added later I think.

File: core/src/test/scala/akka/persistence/jdbc/serialization/StoreOnlySerializableMessagesTest.scala
Comment (patriknw) [reply core/src/test/scala/akka/persistence/jdbc/serialization/StoreOnlySerializableMessagesTest.scala:59]: Such a waste to spend time on flaky tests in slow travis or similar env. It will not test better by a short timeout here.

File: core/src/main/resources/schema/oracle/oracle-schema.sql
Comment (octonato) [reply core/src/main/resources/schema/oracle/oracle-schema.sql:21]: I was thinking about what I just said here. 

Maybe we should indeed split the scripts and have: `drop`, `create`, `dropAndCreate` and `applyScript`. The later being the generic one.

File: core/src/test/scala/akka/persistence/jdbc/journal/JdbcJournalPerfSpec.scala
Comment (octonato) [reply core/src/test/scala/akka/persistence/jdbc/journal/JdbcJournalPerfSpec.scala:54]: I went dig into akka-projections to remember why we did so. Here the comment we have there

```scala
// don't use futureValue (patience) here because it can take a while to start the test container
```

If I remember correctly, creating the DB can take much longer than other async in tests and we didn't want to increase the patience config because of that. So, for the specific case of creating DB we start to use the general Scala API.

File: core/src/main/scala/akka/persistence/jdbc/testkit/internal/SchemaUtilsImpl.scala
Comment (octonato) [inline core/src/main/scala/akka/persistence/jdbc/testkit/internal/SchemaUtilsImpl.scala:60]: @ennru, at the end I did what you suggested. One can pass the whole script in a String and it gets applied. 

The thing is, if you want to load something from the classpath, better to read the file into a string and pass to this method. 

Note that the `applyScript` is the power-user variant, you are on your own to pass all the bits.

**Review by octonato — COMMENTED** [review]
I applied almost all suggestion, plus some other clean-ups:

* use `()` to indicate side-effect
* log in debug in case of errors
* package `testkit`
* `internal.SchemaUtilsImpl` to not create logging confusion
* split scripts into `drop` and `create`
* `applyScript` not dependent on file classpath

**Comment by octonato** [comment]
One more thing, not all schema file has a proper `IF NOT EXISTS`. I need to search more about it, but it seems that this can different a lot.

File: docs/src/main/paradox/index.md
Comment (ennru) [inline docs/src/main/paradox/index.md:83]:
```suggestion
ActorSystem actorSystem = ActorSystem.create("example");
CompletionStage<Done> done = SchemaUtils.dropAndCreateSchema(actorSystem);
```

**Review by ennru — APPROVED** [review]
This grew a bit over these days, good job @octonato!

File: core/src/test/resources/logback-test.xml
Comment (octonato) [inline core/src/test/resources/logback-test.xml:40]:
```suggestion
    <logger name="akka.persistence.jdbc.testkit.scaladsl.SchemaUtils" level="DEBUG"/>
```

File: core/src/main/scala/akka/persistence/jdbc/testkit/javadsl/SchemaUtils.scala
Comment (octonato) [reply core/src/main/scala/akka/persistence/jdbc/testkit/javadsl/SchemaUtils.scala:94]: why?

Comment (ennru) [reply core/src/main/scala/akka/persistence/jdbc/testkit/javadsl/SchemaUtils.scala:94]: Having docs without content doesn't help the user. So either write something useful or remove them.

Comment (octonato) [reply core/src/main/scala/akka/persistence/jdbc/testkit/javadsl/SchemaUtils.scala:94]: oh, you mean the `@return`? I thought you were talking about the whole docs.

Comment (ennru) [reply core/src/main/scala/akka/persistence/jdbc/testkit/javadsl/SchemaUtils.scala:94]: The `@return` without docs should go, and `@param actorSystem` should either go or say `an actor system, classic och new API`

Comment (octonato) [reply core/src/main/scala/akka/persistence/jdbc/testkit/javadsl/SchemaUtils.scala]: done

**Comment by octonato** [comment]
@ennru. I did one more change here after you last approval. 

See my comment in https://github.com/akka/akka-platform-guide/pull/422#issuecomment-745551261

In a nutshell, I removed the `dropAndCreateSchema`. Instead, we should use `dropIfExists` and `createIfNotExists`. This reflects better what it happens behind. 

I will adapt projection to have the same.

**Comment by octonato** [comment]
> I will adapt projection to have the same.

Done. See [`e500e7e` (#490)](https://github.com/akka/akka-projection/pull/490/commits/e500e7eebdf7f62f3166c99ac5a842beb1452677)

**Review by ignasi35 — APPROVED** [review]

**Comment by octonato** [comment]
@ennru, this should pass on first run (famous last words).

`DropCreate` is re-introduced and now relies on `internal.SchemaUtilsImpl` to do its job. `SchemaUtilsImpl` exposes a little bit more (internal API) in order to allow it. 

Previous changes I did was too aggressive and was causing Oracle tests to fail. That's why `DropCreate` and `Cleaners` are not back.

Main changes are in `DropCreate` and `SchemaUtilsImpl`, the rest is are basically the same.

**Pull request merged**
